Replace remaining usages of php strftime

smg6511 opened the pull request:
### What does it do?
Removes usage of the deprecated `strftime` function, implementing the recently-added `modManagerDateFormatter` class formatting methods in its place.

### Why is it needed?
PHP plans to remove `strftime` in v9.0.

### How to test
There were five areas where `strftime` was replaced: 
1. Output filters (date, strftime, fuzzydate), 
2. Date TV (Input Options Disabled Dates field examples and Output Options Date Format field & examples), 
3. Resource Overview (data controller), 
4. Package Browser (search results _released_ date), and
5.  modFIle, which is limited in its implementation; it's not the primary file system handler (which is flysystem based). The one usage I played with was the Export functionality for Form Customization sets. Here I'm not sure that the date is ever used anywhere, so this seems to be more a matter of verifying the Export functionality itself is working as expected.

Note that a new static method was added to `modManagerDateFormatter` to convert existing strftime formats (e.g., `'%A, %d %B %Y'`) to their datetime equivalents (e.g., `'l, d F Y'`) for BC.

That said, just play around in these areas and ensure all works as expected.

### Related issue(s)/PR(s)
Resolves the issue raised in #15864. The subject of moving to the `Intl` extension and its `IntlDateFormatter` methods discussed in that same issue should probably be a separate feature proposal, which would ultimately be implemented within MODX's `modManagerDateFormatter`.

smg6511 committed:
Update modFile and modFileHandler

Replace strftime with modManagerDateFormatter

smg6511 committed:
Format modFile and modFileHandler

Code quality fixes

smg6511 committed:
Update modFile.php

Update doc blocks, ref datetime instead of strftime format

smg6511 committed:
Update modFileHandler.php

Make how formatter is instantiated consistent with other usages

smg6511 committed:
Update data.class.php

Replace strftime with formatter

smg6511 committed:
Update modTransportProvider.php

Replace strftime with formatter

smg6511 committed:
Update modTransportProvider.php

Code style/quality fixes

smg6511 committed:
TV-related updates

Replace strftime with formatter and add backward compat support for existing TVs using strftime formatting strings (new formatter static method to convert these to their datetime equivalents)

smg6511 committed:
Update date.class.php

Code style/quality fixes

smg6511 committed:
TV-related follow up

Fix default format in date and update format conflict prevention between string and date output filter types

smg6511 committed:
Update modOutputFilter.php

Replace strftime with formatter

smg6511 committed:
Update modOutputFilter.php

Code style/quality fixes

halftrainedharry commented:
> The subject of moving to the `Intl` extension and its `IntlDateFormatter` methods discussed in that same issue should probably be a separate feature proposal, which would ultimately be implemented within MODX's `modManagerDateFormatter`.

So what exactly is the plan here?
I'm a bit concerned about the PR is this current form. We have a lot of non-english sites and it's quite common to output the month (or day of the week) as a "full textual representation". E.g. like this:

```
[[*my_date_tv:strtotime:date=`%d. %B %Y`]]
```
With this PR, the **displayed output changes**, and the name of the month is always in english. Which shouldn't be the case.

smg6511 commented:
@halftrainedharry - Good point, I was kind of "running with it" given that everything else had been changed away from `strftime`. The couple places I need to reconsider are the Output Filter and TV Output (Date) Option since their used to format public-facing content. Let me look at integrating `Intl` in now; does anyone object to doing that? If we're not ready to fully commit to `Intl` I could add some checks in and still use strftime if `Intl` is unavailable and php < 9. Thoughts?

smg6511 committed:
Support localized frontend dates

Progress commit (some work still tbd): Backs out changes to output filters and date tv props that made them non-localized by using php datetime. Adds ability to convert strftime to Intl date.

smg6511 commented:
Note that the latest commit reflects work-in-progress to address the need for localized date display. At this point I'm looking more for opinion on where I'm going with this, structure/stragegy-wise. 

I'm thinking of adding a system setting allowing for explicit adoption of `Intl` now, with options for: [1] backend (core manager areas), [2] frontend (tv output), or [3] both. Also I think a new `idate` output filter would be useful (but maybe more appropriate to put in a follow-up PR)—this would be a date filter that exclusively uses `IntlDateFormatter`.

The backend would be much easier for users to move to, as they'd only have to change one formatting setting (`manager_date_format`) to an `Intl` compatible one. The frontend change _could_ be more involved, where folks might have multiple places they've set a `strftime` (or maybe datetime, but I don't think so?) formatting string. That said, in this update I baked in an auto conversion for when `strftime` is no longer available (so far it's looking like php 9 is when it'll be removed); that might be leveraged to enable auto-converted frontend implementation, allowing users to incrementally change their `strftime` patterns to `Intl` ones. 

I don't believe Extras are a factor here (?), as ones that use a date in one way or another have their own formatting settings and implementations.

Ibochkarev commented:
## Code Review

**Summary:** The PR correctly addresses the deprecation of `strftime` (PHP 9) and adds Intl-based localization for frontend dates. The approach with `modFrontendDateFormatter`, `modDateFormatConverter`, and strftime→Intl/datetime mapping is sound.

**Strengths:**
- Replaces `strftime` with `modManagerDateFormatter` / `modFrontendDateFormatter` across output filters, TV configs, modFile, modTransportProvider, and resource data.
- Adds backward-compatible conversion of strftime format strings to Intl/datetime equivalents.
- Addresses @halftrainedharry's concern about localized month/day names for non-English sites.
- New `tabs2spaces` filter and lexicon loading for date-related filters are useful additions.

**Issues to fix before merge:**

1. **Debug code in `modFrontendDateFormatter`** — Constructor and `format()` log to `modX::LOG_LEVEL_ERROR` on every call, which will flood error logs. Remove the `$msg` logging blocks.

2. **`$_SESSION['manager_language']` in frontend context** — `modFrontendDateFormatter` is used for frontend output filters; `$_SESSION['manager_language']` may not exist there. Use `cultureKey` or a safe fallback instead.

3. **`modStrftimeToIntlConverter`** — Contains `echo 'CONV working...';` (debug output). Remove before merge. The class also appears to be a stub (`return false`); confirm it is either implemented or not yet wired in.

4. **Date filter input parsing** — The original logic used `strtotime($output)` for non-timestamp strings. `parseValue()` in the parent only accepts timestamps or `Y-m-d H:i:s` MySQL format. Date strings like `2024-01-15` or `January 15, 2024` may no longer parse correctly. Consider preserving `strtotime()` fallback for non-timestamp strings before passing to the formatter.

**Verdict:** The direction is good and the structure is solid. Remove debug code and fix the frontend/session and parsing edge cases before marking ready for review.

modxcommunity commented:
This pull request has been mentioned on **MODX Community**. There might be relevant details there:

https://community.modx.com/t/strftime-errors/9044/6
